internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,214 @@
+		clusterDeploymentRefErr := newKubeAPIError(
+			errors.Wrapf(
+				err,
+				fmt.Sprintf("failed to find clusterDeployment with name %s in namespace %s",
+					installEnv.Spec.ClusterRef.Name, installEnv.Spec.ClusterRef.Namespace)),
+			k8serrors.IsNotFound(err),
+		)
+		return r.updateStatusAndReturnResult(ctx, installEnv, nil, imageState, clusterDeploymentRefErr)
+	}
+
+	c, err := r.Installer.GetClusterByKubeKey(types.NamespacedName{
+		Name:      clusterDeployment.Name,
+		Namespace: clusterDeployment.Namespace,
+	})
+
+	if gorm.IsRecordNotFoundError(err) {
+		inventoryErr = common.NewApiError(http.StatusNotFound, err)
+
+	} else {
+		isoParams := installer.GenerateClusterISOParams{
+			ClusterID:         *c.ID,
+			ImageCreateParams: &models.ImageCreateParams{},
+		}
+		if len(installEnv.Spec.SSHAuthorizedKeys) > 0 {
+			isoParams.ImageCreateParams.SSHPublicKey = installEnv.Spec.SSHAuthorizedKeys[0]
+		}
+		updatedCluster, inventoryErr = r.Installer.GenerateClusterISOInternal(ctx, isoParams)
+	}
+
+	if inventoryErr == nil {
+		imageInfo = updatedCluster.ImageInfo
+		imageState.State = ImageBeingCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+	}
+
+	return r.updateStatusAndReturnResult(ctx, installEnv, imageInfo, imageState, inventoryErr)
+}
+
+func (r *InstallEnvReconciler) updateStatusAndReturnResult(
+	ctx context.Context,
+	installEnv *adiiov1alpha1.InstallEnv,
+	imageInfo *models.ImageInfo,
+	imageState ImageState,
+	err error) (ctrl.Result, error) {
+
+	var res ctrl.Result
+
+	if imageBeingCreated(err) {
+		imageState.State = ImageCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+		// Clear up the error stateInfo while installEnv is being created
+		err = nil
+		r.Log.Infof("Image %s being prepared for cluster %s stateInfo: %s",
+			installEnv.Name, installEnv.ClusterName, imageState.StateInfo)
+	} else if isClientError(err) {
+		imageState.State = ImageCreationError // TODO before merge: should we do this?
+		imageState.StateInfo += ": " + err.Error()
+	} else if err != nil {
+		imageState.State = ImageCreationError
+		imageState.StateInfo += ": internal error"
+		res.Requeue = true
+	}
+	r.setStateAndStateInfo(adiiov1alpha1.ImageProgressCondition, imageState, installEnv)
+
+	if err == nil && imageInfo != nil {
+		installEnv.Status.ISODownloadURL = imageInfo.DownloadURL
+	} else if err != nil {
+		r.Log.WithError(err).Error("installEnv reconcile failed")
+	}
+
+	if updateErr := r.Status().Update(ctx, installEnv); updateErr != nil {
+		r.Log.WithError(updateErr).Error("failed to update installEnv status")
+		res.Requeue = true
+		return res, nil
+	}
+
+	return res, nil
+}
+
+func (r *InstallEnvReconciler) setStateAndStateInfo(conditionType adiiov1alpha1.InstallEnvConditionType,
+	imageState ImageState, installEnv *adiiov1alpha1.InstallEnv) {
+	r.setCondition(adiiov1alpha1.InstallEnvCondition{
+		Type:               conditionType,
+		Status:             corev1.ConditionUnknown,
```

Comment:
I don't see how this ever transitions out of the `Unknown` condition. Am I missing it?

Reply:
@nmagnezi i think that there are two options
1. Make it optional, after all reason and message will show the actual state. 
2. True if there is no error, else false. by no error i mean status is created.

Reply:
I picked option 2, please take a look